mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/transformation/mpsToModelix/initial/ModelSynchronizer.kt

```diff
@@ -62,30 +65,53 @@ class ModelSynchronizer(private val branch: IBranch, postponeReferenceResolution
 
     fun addModel(model: SModelBase) =
         syncQueue.enqueue(linkedSetOf(SyncLock.MODELIX_WRITE, SyncLock.MPS_READ), SyncDirection.MPS_TO_MODELIX) {
-            val moduleModelixId = nodeMap[model.module]!!
-            val models = BuiltinLanguages.MPSRepositoryConcepts.Module.models
-
+            val parentModule = model.module!!
+            val moduleModelixId = nodeMap[parentModule]!!
             val cloudModule = branch.getNode(moduleModelixId)
-            val cloudModel = cloudModule.addNewChild(models, -1, BuiltinLanguages.MPSRepositoryConcepts.Model)
+            val childLink = BuiltinLanguages.MPSRepositoryConcepts.Module.models
+
+            // duplicate check
+            val modelId = model.modelId.toString()
+            val modelExists = cloudModule.getChildren(childLink)
+                .firstOrNull { modelId == it.getPropertyValue(BuiltinLanguages.MPSRepositoryConcepts.Model.id) } != null
+            if (modelExists) {
+                if (nodeMap.isMappedToModelix(model)) {
+                    return@enqueue ItemAlreadySynchronizer(model)
+                } else {
+                    throw Exception("Model ${model.name} in Module ${parentModule.moduleName} already exists on the server, therefore it and its parent module will not be synchronized completely. Remove the parent module from the project and synchronize it from the server instead.")
+                }
+            }
 
+            val cloudModel = cloudModule.addNewChild(childLink, -1, BuiltinLanguages.MPSRepositoryConcepts.Model)
             nodeMap.put(model, cloudModel.nodeIdAsLong())
-
             synchronizeModelProperties(cloudModel, model)
 
             // synchronize root nodes
             model.rootNodes.waitForCompletionOfEachTask { nodeSynchronizer.addNode(it) }
-        }.continueWith(linkedSetOf(SyncLock.MPS_READ), SyncDirection.MPS_TO_MODELIX) {
+        }.continueWith(linkedSetOf(SyncLock.MPS_READ), SyncDirection.MPS_TO_MODELIX) { statusToken ->
+            if (statusToken is ItemAlreadySynchronizer) {
+                return@continueWith statusToken
+            }
+
             // synchronize model imports
             model.modelImports.waitForCompletionOfEachTask { addModelImport(model, it) }
-        }.continueWith(linkedSetOf(SyncLock.MPS_READ), SyncDirection.MPS_TO_MODELIX) {
+        }.continueWith(linkedSetOf(SyncLock.MPS_READ), SyncDirection.MPS_TO_MODELIX) { statusToken ->
+            if (statusToken is ItemAlreadySynchronizer) {
```

Comment:
How would coroutines help in this case? The problem is, we have tasks that need certain locks (must run a certain threads). These tasks must wait for each other if the lock is unavailable. However, they should not block the thread on which they are running if the lock they are waiting for is not available.

That is what the SyncQueue with the ContinuableSyncTasks and the CompletableFutures prevent. SyncQueue takes care of running the ContinuableSyncTasks with the correct locks. The CompletableFuture makes the execution of the tasks "suspendable"/continuable, in a sense that we pause the execution of the task if we have to wait for some other tasks to complete (and those other tasks need the locks that we are holding). However, in this case the continuation is a new ContinuableSyncTask (e.g. in a following `.continueWith` block) but at least the continuations can be followed threw the CompletableFuture chain.

Reply:
```
For locks, we can use mutexes.
Coroutines automatically continuing execution after mutexes are free
```

Lock in this context means, the action has to be run on a certain thread. [Here](https://github.com/modelix/modelix.mps-plugins/blob/feature/modelix-755/mps-sync-plugin-lib/src/main/kotlin/org/modelix/mps/sync/tasks/SyncQueue.kt#L151) is where the "locks" are acquired. You can click through the call chain to see what I mean.

> But this is definitely something out of scope for this PR

Thanks, I'm much relieved. One problem less to take care of in this PR. 😅 Changing the ThreadPool-SyncTask-CompletableFuture-based architecture to a Coroutine-based one is a really big refactoring since that is the soul of the task execution workflow of the plugin. It's a non-trivial task with many corner-cases that I discovered while implementing it.